test/extended/templates/templateinstance_cross_namespace.go

```diff
@@ -0,0 +1,139 @@
+			},
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		templateinstance := &templateapi.TemplateInstance{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "templateinstance",
+			},
+			Spec: templateapi.TemplateInstanceSpec{
+				Template: templateapi.Template{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "template",
+						Namespace: cli.Namespace(),
+					},
+					Parameters: []templateapi.Parameter{
+						{
+							Name: "NAMESPACE",
+						},
+					},
+				},
+				Secret: kapi.LocalObjectReference{
+					Name: "secret",
+				},
+			},
+		}
+
+		err = templateapi.AddObjectsToTemplate(&templateinstance.Spec.Template, []runtime.Object{
+			// secret in the same namespace
+			&kapi.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "secret1",
+				},
+			},
+			// secret in a different namespace
+			&kapi.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "secret2",
+					Namespace: "${NAMESPACE}",
+				},
+			},
+		}, latest.Versions...)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("creating the templateinstance")
+		_, err = cli.TemplateClient().Template().TemplateInstances(cli.Namespace()).Create(templateinstance)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// wait for templateinstance controller to do its thing
+		err = wait.Poll(time.Second, time.Minute, func() (bool, error) {
+			templateinstance, err = cli.TemplateClient().Template().TemplateInstances(cli.Namespace()).Get(templateinstance.Name, metav1.GetOptions{})
+			if err != nil {
+				return false, err
+			}
+
+			for _, c := range templateinstance.Status.Conditions {
+				if c.Reason == "Failed" && c.Status == kapi.ConditionTrue {
+					return false, fmt.Errorf("failed condition: %s", c.Message)
+				}
+				if c.Reason == "Created" && c.Status == kapi.ConditionTrue {
+					return true, nil
+				}
+			}
+
+			return false, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		_, err = cli.KubeClient().CoreV1().Secrets(cli.Namespace()).Get("secret1", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		_, err = cli.KubeClient().CoreV1().Secrets(cli2.Namespace()).Get("secret2", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("deleting the templateinstance")
+		err = cli.TemplateClient().Template().TemplateInstances(cli.Namespace()).Delete(templateinstance.Name, nil)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// wait for garbage collector to do its thing
+		err = wait.Poll(time.Second, time.Minute, func() (bool, error) {
+			_, err = cli.TemplateClient().Template().TemplateInstances(cli.Namespace()).Get(templateinstance.Name, metav1.GetOptions{})
+			if err == nil || !kerrors.IsNotFound(err) {
+				return false, err
+			}
+
+			_, err = cli.KubeClient().CoreV1().Secrets(cli.Namespace()).Get("secret1", metav1.GetOptions{})
+			if err == nil || !kerrors.IsNotFound(err) {
+				return false, err
+			}
+
+			_, err = cli.KubeClient().CoreV1().Secrets(cli2.Namespace()).Get("secret2", metav1.GetOptions{})
```

Comment:
> @deads2k is this part of how GC works?

This is an accident, not an intent.  You should open an issue upstream and we can decide if we want to enshrine it as expected behavior.  I don't have strong feelings either way since it is UID dependent.